Port run_subprocess to take the command as a list instead of a string.

abadger opened the pull request:
Security hardening: change the run_subprocess() function to take the command as a list instead of a string.

This is complete and ready for review. @Andrew-ang9 and I have been working on this together.

Implements: https://issues.redhat.com/browse/OAMG-6183

abadger commented on convert2rhel/checks.py:
```diff
@@ -243,11 +243,14 @@ def get_rhel_supported_kmods():
+            "repoquery",
+            "--releasever=%s" % system_info.releasever,
+            setopt_arg,
+            repoquery_repoids_args,
```
I think setopt_arg and reppoquery_repoids_args here need to be modified a bit:
* setopt_arg might be set to the empty string: https://github.com/oamg/convert2rhel/pull/389/files#diff-cf19bc90f65d9e26e1b6fa7dcd84a18dfba097f65706e211a4f4ac71b1b467f6R242  I'm not sure if repoquery will ignore that or attempt to do something with it that fails.  It would be better to only add setopt_arg if it exists.
* repoquery_repoids_args is being set to a string with spaces for each repository like this:
  `'--repoid fedora --repoid fedora-updates'`.  We need each of those to be separate entries in the list.

Given those constraints, I think we need to build up the command as a list outside of the run_subprocess() rather than trying to construct the list inside of the function call.  Something like we did in pkghandler: https://github.com/oamg/convert2rhel/pull/389/files#diff-361f5dc14483ec43e37529967acee396da5714dbf6562472ccece5db4bba5e0dR134

abadger replied:
@Andrew-ang9 Could you take a look at fixing this one?

Andrew-ang9 replied:
Yeah I can take care of that!

r0x0d commented:
@abadger, a quick question... Should this PR cover the same security hardening for: https://github.com/oamg/convert2rhel/blob/120ad4f6356f50e861acc15f826132c4c9d0e7e3/convert2rhel/utils.py#L160

As it seems, this command is not used in many places, only here: https://github.com/oamg/convert2rhel/blob/120ad4f6356f50e861acc15f826132c4c9d0e7e3/convert2rhel/utils.py#L452-L472

If this method could benefit from the same process, I guess it could be good to add the same mitigations for this method as well.

r0x0d commented on convert2rhel/pkghandler.py:
```diff
@@ -124,12 +124,14 @@ def call_yum_cmd(
 
     YUM/DNF typically expands the $releasever variable used in repofiles. However it fails to do so after we remove the
-    release packages (centos-release, oraclelinux-release, etc.) and before the redhat-release package is installed.
+    e packages (centos-release, oraclelinux-release, etc.) and before the redhat-release package is installed.
```
```suggestion
    packages (centos-release, oraclelinux-release, etc.) and before the redhat-release package is installed.
```

r0x0d commented on convert2rhel/subscription.py:
```diff
@@ -126,13 +126,13 @@ def register_system():
         # -> organization required
         loggerinst.info("    ... activation key detected: %s" % tool_opts.activation_key)
-        registration_cmd += ' --activationkey="%s"' % tool_opts.activation_key
+        registration_cmd.append("--activationkey=%s" % tool_opts.activation_key)
```
The parameter is missing the quotes for %s

```suggestion
        registration_cmd.append('--activationkey="%s"' % tool_opts.activation_key)
```

abadger replied:
I don't think quotes are needed.  When we pass in a string and then split it into arguments using `shlex.split()` we have to be careful that anything that could contain spaces is enclosed in quotes.  Otherwise `shlex.split()` will split it into two arguments.  When we pass a list of arguments, instead, no additional splitting will occur, each element of the list will be a single argument:
``` python
>>> import subprocess, shlex
>>> open('test case', 'w').close()
>>> # Oops, we forgot to quote the filename
>>> subprocess.Popen(shlex.split('cat test case')).wait()
cat: test: No such file or directory
cat: case: No such file or directory
1
>>> # This is how we have to construct the command as a string:
>>> subprocess.Popen(shlex.split('cat "test case"')).wait()
0
>>> # When we use a list, we don't have to quote:
>>> subprocess.Popen(['cat', 'test case']).wait()
0
```
Not having to quote is part of why this is a more secure API for this function.  It's easy to forget to quote something when constructing a string to pass in (especially when dealing with variables). The code will then work until someone constructs a string with a space in it (possibly maliciously).  Passing a list instead means that we do not have to remember to quote.

bocekm replied:
I agree that the quotes are not needed. Frankly I needed to try it out myself with the _=_ sign to verify it 🤫 The following Popen call:
```python
>>> subprocess.Popen(['subscription-manager', 'register', '--serverurl=host.example.com; rm file']).wait()
Unable to reach the server at host.example.com; rm file:443/subscription
69
```
... is the same as if running the following in shell:
```shell
# subscription-manager register "--serverurl=host.example.com; rm file"
Unable to reach the server at host.example.com; rm file:443/subscription
```

The following code construct would not make me think twice about the functionality. But that would change the command called and that's I believe not the intent of this PR.
```suggestion
        registration_cmd.extend(["--activationkey", "%s" % tool_opts.activation_key])
```

r0x0d replied:
Oh, that's cool! Thought that we had to preserve the quotes for these cases (Imagined that this could cause some unreliable calls to subscription-manager). Either way, the other comments about quotes can be ignored safely then!

r0x0d commented on convert2rhel/subscription.py:
```diff
@@ -153,7 +153,7 @@ def get_registration_cmd():
                 loggerinst.info("Username: " + username)
             password = utils.prompt_user("Password: ", password=True)
-        registration_cmd += ' --username=%s --password="%s"' % (username, password)
+        registration_cmd.extend(["--username=%s" % username, "--password=%s" % password])
```
```suggestion
        registration_cmd.extend(["--username=%s" % username, '--password="%s"' % password])
```

r0x0d commented on convert2rhel/subscription.py:
```diff
@@ -167,16 +167,16 @@ def get_registration_cmd():
     if tool_opts.serverurl:
         loggerinst.debug("    ... using custom RHSM URL")
-        registration_cmd += ' --serverurl="%s"' % tool_opts.serverurl
+        registration_cmd.append("--serverurl=%s" % tool_opts.serverurl)
```
```suggestion
        registration_cmd.append('--serverurl="%s"' % tool_opts.serverurl)
```

r0x0d reviewed (changes requested):
The changes look good! Just a minor comment on a few changes that looked different from the original ones. 

There is a `newfile.txt` in the PR, maybe it was created by mistake.

abadger commented on convert2rhel/utils.py:
```diff
@@ -119,24 +119,22 @@ def restart_system():
+    The cmd is specified as a list starting with the command and followed by a list of arguments.
+    Example: ["dnf", "repoquery", "kernel"]
     """
     if print_cmd:
```
If we do, then we probably want to check whether this is a string rather than checking if it is not a list to preserve duck-typing (subprocess.Popen should work with anything that is a Sequence [a tuple, for instance.])

A question, though:
* Are we guarding against str being passed in because str used to be passed in?
  * If so, we probably want to limit the time that we perform this check.
  * If not, what is the difference between this function and other functions which take lists and would error if given a string?

r0x0d replied:
I think you're right, we should compare against `str` and not `list`, which makes more sense. 

Regarding the question, I thought about checking against this because of what you mentioned in your gist. Someone could forget to use `shlex.quote()` when passing an `str` to our `run_subprocess()` and thus, we would be in the same spot as of right now. 

Let me know what you think about that, it's just a suggestion (don't know either if I understood correctly what I read 😅)

SpyTec commented:
/packit build

abadger commented:
@r0x0d I've incorporated your suggestions and modified  run_cmd_in_pty().  I'll let @Andrew-ang9 modify setopt_arg and reppoquery_repoids_args in checks.py and then this should be ready for another review.

abadger commented:
> 

I think that all of the things you've mentioned have now been fixed.  Let me know if there's more problems!

(And you were right about newfile... I meant to delete it after we tested that andrew could push to my repo).

abadger commented:
/pakit build

abadger commented:
/packit build

r0x0d commented on convert2rhel/pkghandler.py:
```diff
@@ -158,10 +160,10 @@ def call_yum_cmd(
-        cmd += " --enablerepo=%s" % repo
+        cmd.append("--enablerepo=%s" % repo)
 
     if args:
```
I think this check is not entirely necessary here, at the beginning we are already saying that if the `args is None`, then `args = []`.

What do you think, @abadger?

abadger replied:
I agree!  Removing it now.

abadger commented:
Sorry, I messed up my rebase.  Give me a moment to fix that.

abadger commented:
Alright, fixed.

kokesak commented:
@SpyTec yes, I was just checking it. I wil try to re-run it again. If it fails again I think it's ok to merge

abadger commented:
Thanks for the reviews @r0x0d @SpyTec and @kokesak !
